runtime/lua/vim/lsp/buf.lua

```diff
@@ -571,35 +581,48 @@ end
 --- Requests code actions from all clients and calls the handler exactly once
 --- with all aggregated results
 ---@private
-local function code_action_request(params)
+local function code_action_request(params, opts)
   local bufnr = vim.api.nvim_get_current_buf()
   local method = 'textDocument/codeAction'
   vim.lsp.buf_request_all(bufnr, method, params, function(results)
-    on_code_action_results(results, { bufnr = bufnr, method = method, params = params })
+    ctx = { bufnr = bufnr, method = method, params = params, opts = opts}
+    on_code_action_results(results, ctx)
   end)
 end
 
 --- Selects a code action available at the current
 --- cursor position.
 ---
----@param context table|nil `CodeActionContext` of the LSP specification:
----               - diagnostics: (table|nil)
----                             LSP `Diagnostic[]`. Inferred from the current
----                             position if not provided.
----               - only: (string|nil)
----                      LSP `CodeActionKind` used to filter the code actions.
----                      Most language servers support values like `refactor`
----                      or `quickfix`.
+---@param opts table|nil Optional table which holds the following optional fields:
+---    - context (table|nil):
+---        Corresponds to `CodeActionContext` of the LSP specification:
+---          - diagnostics: (table|nil)
+---                        LSP `Diagnostic[]`. Inferred from the current
+---                        position if not provided.
+---          - only: (string|nil)
+---                 LSP `CodeActionKind` used to filter the code actions.
+---                 Most language servers support values like `refactor`
+---                 or `quickfix`.
+---    - selection_filter: (function|nil)
+---                       Predicate function taking an `CodeAction` and returning a boolean.
+---                       When the filter results in just one remaining action, that action
+---                       is applied automatically.
```

Comment:
What do you think of de-coupling the auto-select into a dedicated option - that if there is only one result it is automatically applied?